data/headset_dataset.py

```diff
@@ -0,0 +1,256 @@
+        segment, frame = self.framelist["seg_id"][idx], self.framelist["frame_id"][idx]
+        segment_name = segment.split("-")[0]
+        if self.load_latent_code:
+            entry_name = f"{self.capture.sid}-{segment}_{segment_name}_{frame}"
+            try:
+                latent_code = pickle.loads(
+                    ZipPath(self.latent_code_dir / "rosetta_correspondences.zip", entry_name).read_bytes()
+                )["expression"]
+            except FileNotFoundError:
+                return None
+        else:
+            print("Using placeholder for GT code.")
+            latent_code = np.zeros((self.self.latent_code_dim,))
+
+        return {
+            "headset_cam_img": self.get_img_by_segment_and_frame(segment, frame),
+            "cond_headset_cam_img": self.neutral_img[:, None],
+            "gt_latent_code": torch.from_numpy(latent_code),
+            "index": {"ident": self.capture.sid, "segment": segment_name, "frame": frame},
+        }
+
+    def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
+        return self.get_assets_by_id(idx)
+
+    def __len__(self) -> int:
+        return len(self.framelist)
+
+    def get_random_frame(self):
+        idx = np.random.randint(0, len(self.framelist))
+        return self.get_assets_by_id(idx)
+
+    def get_img_size(self) -> Tuple[int, int]:
+        return (self.height, self.width)
+
+
+class MultiCaptureDataset(torch.utils.data.IterableDataset):
+    """
+    Dataset with headset images for multiple captures
+
+    Args:
+        captures: A list of unique identifiers of the headset captures in this dataset
+        directories: A list of the directory paths for each capture
+        downsample: Downsampling factor to use when loading images
+        time_downsample: Time downsampling factor to use when loading images
+        cameras_specified: A list of the headset camera names to load
+    """
+
+    def __init__(
+        self,
+        captures: List[HeadsetCapture],
+        directories: List[str],
+        latent_code_directories: Optional[List[Union[str, Optional[str]]]],
+        downsample: int = 4,
+        time_downsample: int = 1,  # no time downsample
+        cameras_specified: List[str] = ["left-eye-atl-temporal", "right-eye-atl-temporal", "left-mouth", "right-mouth"],
+    ):
+        self.identities = captures
+        if latent_code_directories is None:
+            latent_code_directories = [None] * len(captures)
+        self.datasets = [
+            SingleCaptureDataset(cap, cap_dir, code_dir, downsample, time_downsample, cameras_specified)
+            for cap, cap_dir, code_dir in zip(captures, directories, latent_code_directories)
+        ]
+
+    def __iter__(self):
+        while True:
+            dataset_idx = np.random.randint(0, len(self.datasets))
+            # dataset_idx = torch.randint(0, len(self.datasets), (1,)).item()
+            dataset = self.datasets[dataset_idx]
+            yield dataset.get_random_frame()
```

Comment:
This is doing stratified sampling, which ideally should be done by a `DatasetSampler` on top of a dataset with random access. Could we at least leave a note with a TODO to improve this at a later stage?

Reply:
Currently the dataset is a subclass of the IterableDataset, which I believe doesn't support using samplers. Specifically, we want an infinite looping over the data, rather than epoch-based training.

For that reason, an alternative may be to re-write the dataset as a map-style dataset and then add a sampler with random access and replacement when we define dataloaders. If you think that makes sense, I'll leave a TODO for that path. Please let me know!

Reply:
It makes sense that iterables don't support samplers -- you'd need a random access dataset for that. We can also make an infinite sampler for the non-stop training over a finite number of iterations. A TODO is probably fine for now.